### PR TITLE
Implement NamedMultiReservableCurrency and NamedBasicReservableCurrency in orml_currencies

### PR DESCRIPTION
**Pull request by Chralt98**
Hey, this is related to #742. 

It's the first try and it's not ready yet.

### CODE REVIEW

**Comment by xlc** [comment]
Ok. Simply add the trait bound makes no sense (that's what I read from the issue) but actually implement `NamedMultiReservableCurrency` is useful.

**Comment by Chralt98** [comment]
@xlc I don't get why the cargo fmt isn't correct. I formatted it multiple times.

**Comment by Chralt98** [comment]
I let you finish the work. I think the most is done.

File: currencies/src/lib.rs [hunk 1]
```diff
@@ -84,11 +85,13 @@ pub mod module {
 		type MultiCurrency: TransferAll<Self::AccountId>
 			+ MultiCurrencyExtended<Self::AccountId>
 			+ MultiLockableCurrency<Self::AccountId>
-			+ MultiReservableCurrency<Self::AccountId>;
+			+ MultiReservableCurrency<Self::AccountId>
+			+ NamedMultiReservableCurrency<Self::AccountId, ReserveIdentifier = [u8; 8]>;
```
Comment (zqhxuyuan) [inline currencies/src/lib.rs:89]: It's better to define a ReserveIdentifier type config here.

Comment (xlc) [reply currencies/src/lib.rs:89]: There isn't a real need to define it. Just leave it out and use `ReserveIdentifierOf` when you need it.

`type ReserveIdentifierOf<T> = <<T as Config>::MultiCurrency as NamedMultiReservableCurrency <<T as frame_system::Config>::AccountId>>:: ReserveIdentifier;`

**Comment by Chralt98** [comment]
I don't really know if [this](https://github.com/open-web3-stack/open-runtime-module-library/blob/c76ec675ea7f3334a0dffbb4d195f49147c34a4a/currencies/src/lib.rs#L779) might be a problem in the future. I mean we got a `ReserveIdentifier` for `orml_tokens` AND another `ReserveIdentifier` for `pallet_balances`.

PS: I hate this cargo fmt problem here in this repo...

**Comment by xlc** [comment]
Update your cargo to a recent version should resolve the fmt issue. This is one used by CI
https://github.com/open-web3-stack/open-runtime-module-library/blob/39057dae9c277e34a8e1c61abf2148158199e687/.github/workflows/test.yml#L36

**Comment by xlc** [comment]
Multiple `ReserveIdentifier` is not an issue. You can require them to be the same one if you need. Similar to how the `Balance` type for pallet-balances and orml-tokens are handled.

**Comment by Chralt98** [comment]
Thanks @xlc ! If there is anything to correct in my pull request, I am open to correct it after a review.

File: currencies/src/lib.rs [hunk 2]
```diff
@@ -619,7 +738,7 @@ where
 impl<T, AccountId, Currency, Amount, Moment> BasicReservableCurrency<AccountId>
 	for BasicCurrencyAdapter<T, Currency, Amount, Moment>
 where
-	Currency: PalletReservableCurrency<AccountId>,
+	Currency: PalletNamedReservableCurrency<AccountId>,
```
Comment (xlc) [inline currencies/src/lib.rs:741]: why change this?

Comment (Chralt98) [reply currencies/src/lib.rs]: Sorry, this was my bad.

File: currencies/src/lib.rs [hunk 3]
```diff
@@ -368,6 +375,75 @@ impl<T: Config> MultiReservableCurrency<T::AccountId> for Pallet<T> {
 	}
 }
 
+impl<T: Config> NamedMultiReservableCurrency<T::AccountId> for Pallet<T> {
+	type ReserveIdentifier = ReserveIdentifierOf<T>;
```
Comment (xlc) [inline currencies/src/lib.rs:379]:
```suggestion
	type ReserveIdentifier = ReserveIdentifierOf<T>;

```

**Comment by zqhxuyuan** [comment]
LGTM except missing testcase, can you do this here or maybe in another PR? @Chralt98

**Comment by Chralt98** [comment]
@zqhxuyuan This depends on, what would be faster for a merge (Zeitgeist requires this functionality as fast as possible). I would propose, that it should get another pull request. But please verify the code before merge again :-D. I don't like to impact many chains to loose money. I would try to add test cases later in the day in another pull request. But I would be happy, if you could spend time on this too.

**Review by zqhxuyuan — APPROVED** [review]
LGTM

**Comment by zqhxuyuan** [comment]
It's better to add `can_reserve` method too, you can do this on later PR. @Chralt98

**Comment by Chralt98** [comment]
@zqhxuyuan You mean `can_reserve_named` for `NamedBasicReservableCurrency` and for `NamedMultiReservableCurrency` then? But this is not implemented for `orml_tokens`. Or what do you mean?

**Comment by zqhxuyuan** [comment]
ok, my bad. the `can_reseve` is not needed for named implementation.

**Pull request merged**